lib/fauna_model/base.rb

```diff
@@ -0,0 +1,269 @@
+          model.send(:from_resource, resource)
+        end
+
+        def create(params = {})
+          model = new(params)
+          result = model.save
+          if result
+            result
+          else
+            model
+          end
+        end
+
+        def create!(params = {})
+          model = new(params)
+          model.save!
+          model
+        end
+
+        def find(identifier)
+          identifier = Fauna::Ref.new(identifier) if identifier.is_a? String
+
+          from_fauna(Fauna::Context.query(Fauna::Query.get(identifier)))
+        end
+
+        def find_by_id(id)
+          id = Fauna::Ref.new("#{fauna_class}/#{id}") unless id.is_a? Fauna::Ref
+
+          from_fauna(Fauna::Context.query(Fauna::Query.get(id)))
+        end
+      end
+
+      def initialize(params = {})
+        @original = {}
+        init_state
+        apply_params params
+      end
+
+      def new_record?
+        ref.nil?
+      end
+
+      def deleted?
+        @deleted
+      end
+
+      def persisted?
+        !new_record? && Model.calculate_diff(@original, @current).empty?
+      end
+
+      def id
+        ref.value.split('/').last unless ref.nil?
+      end
+
+      def save(validate = true)
+        return false if validate && invalid?
+
+        if new_record?
+          self.class.from_fauna(Fauna::Context.query(create_query))
+        else
+          self.class.from_fauna(Fauna::Context.query(update_query))
+        end
+      end
+
+      def save!(validate = true)
+        fail InvalidInstance.new('Invalid instance data') if validate && invalid?
+
+        old_changes = changes
+
+        if new_record?
+          result = from_resource(Fauna::Context.query(create_query))
+        else
+          result = from_resource(Fauna::Context.query(update_query))
+        end
+
+        @previous_changes = old_changes
+
+        result
+      end
+
+      def update(params = {})
+        model = copy
+        model.send(:apply_params, params)
+        model.save
+      end
+
+      def update!(params = {})
+        apply_params(params)
+        save!
+      end
+
+      def destroy
+        return false if new_record?
+        Fauna::Context.query(delete_query)
+        @deleted = true
+      end
+
+    private
+
+      def create_query
+        return nil unless new_record?
+        Fauna::Query.create(self.class.fauna_class, Fauna::Query.quote(query_params))
+      end
+
+      def replace_query
+        return nil if new_record?
+        Fauna::Query.replace(@current['ref'], Fauna::Query.quote(query_params))
+      end
+
+      def update_query
+        return nil if new_record?
+        Fauna::Query.update(@current['ref'], Fauna::Query.quote(Model.calculate_diff(@original, @current)))
+      end
+
+      def delete_query
+        return nil if new_record?
+        Fauna::Query.delete(ref)
+      end
+
+      def query_params
+        params = Model.hash_dup(@current)
+
+        # Remove unsettable fields
```

Comment:
Doesn't need to be a setter for read only. The model stores the whole resource, so we always have to filter it out. As for write once, the setter is always there.